generators/app/templates/src/StarterKit/ServiceAgents/_base/AgentBase.cs

```diff
@@ -1,127 +1,238 @@
+using System.Text;
 using System.Threading.Tasks;
+using Digipolis.Errors;
+using Digipolis.Errors.Exceptions;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+using Newtonsoft.Json;
+using StarterKit.ServiceAgents._base.Helper;
+using StarterKit.ServiceAgents._base.Settings;
+using StarterKit.Shared.Exceptions.Models;
+using BadGatewayException = StarterKit.Shared.Exceptions.Models.BadGatewayException;
+using GatewayTimeoutException = StarterKit.Shared.Exceptions.Models.GatewayTimeoutException;
 
 namespace StarterKit.ServiceAgents._base
 {
-    public abstract class AgentBase<TAgent> where TAgent : class
+  public abstract class AgentBase<TAgent> : IDisposable where TAgent : class
+  {
+    protected HttpClient Client { get; set; }
+    protected readonly AgentSettingsBase Settings;
+
+    private readonly ILogger<TAgent> _logger;
+    private readonly IRequestHeaderHelper _requestHeaderHelper;
+    private readonly JsonSerializerSettings _jsonSerializerSettings;
+
+    protected AgentBase(ILogger<TAgent> logger, HttpClient httpClient, IServiceProvider serviceProvider)
     {
-        protected AgentBase(ILogger<TAgent> logger, HttpClient httpClient)
-        {
-            Logger = logger ?? throw new ArgumentException($"{GetType().Name}.Ctor parameter {nameof(logger)} cannot be null.");
-            HttpClient = httpClient ?? throw new ArgumentException($"{GetType().Name}.Ctor parameter {nameof(httpClient)} cannot be null.");
-        }
+      var settings = serviceProvider.GetRequiredService<IOptions<ServiceAgentSettings>>();
+      if (settings.Value == null) throw new ArgumentNullException(nameof(ServiceAgentSettings), $"{nameof(ServiceAgentSettings)} cannot be null.");
 
-        protected ILogger<TAgent> Logger { get; private set; }
-        protected HttpClient HttpClient { get; set; }
+      Client = httpClient;
+      Settings = settings.Value.GetServiceSettings(typeof(TAgent)?.Name);
+      _logger = logger;
+      _requestHeaderHelper = serviceProvider.GetRequiredService<IRequestHeaderHelper>();
+      _jsonSerializerSettings = serviceProvider.GetRequiredService<IOptions<MvcNewtonsoftJsonOptions>>()?.Value?.SerializerSettings;
+    }
 
-        private readonly HttpCompletionOption _defaultHttpCompletionOption = HttpCompletionOption.ResponseHeadersRead;
+    protected virtual void OnParseJsonErrorException(Exception ex, HttpResponseMessage response) { }
 
-        protected async Task<T> GetAsync<T>(string requestUri)
-        {
-            using (var response = await HttpClient.GetAsync(requestUri, _defaultHttpCompletionOption))
-            {
-                using (var stream = await response.Content.ReadAsStreamAsync())
-                {
-                    return await Utf8Json.JsonSerializer.DeserializeAsync<T>(stream);
-                }
-            }
-        }
+    protected async Task ParseJsonError(HttpResponseMessage response)
```

Comment:
Deze logica zou beter in een delegating handler blijven steken. Sommige API's die aangesproken worden hebben soms een specifiek Error-model ipv het Digipolis.Error-model of hebben een specifieke error-afhandeling nodig. Hiervoor zou je dan specifieke delegating handlers kunnen opstellen per ServiceAgent

Reply:
Zelfde bedenking als Erik.